tilequeue/postgresql.py

```diff
@@ -1,297 +1,66 @@
-        return conn
-
-    def get_conns_for_db(self, dbname):
-        conn_info_with_db = dict(self.conn_info, dbname=dbname)
-        conns = [self._make_conn(conn_info_with_db)
-                 for i in range(self.n_conn_per_db)]
-        with self.lock:
-            assert dbname not in self.db_to_conns, \
-                'already in db_to_conns: %s' % dbname
-            self.db_to_conns[dbname] = conns
-        return conns
-
-    def put_conns_for_db(self, dbname):
-        with self.lock:
-            conns = self.db_to_conns.pop(dbname)
-        for conn in conns:
-            try:
-                conn.close()
+                    self._conns_to_pool[id(conn)] = pool_to_use
+                    conns.append(conn)
+                assert len(conns) == n_conns, \
+                    "Couldn't collect enough connections"
             except:
-                pass
-
-    def closeall(self):
-        with self.lock:
-            for dbname, conns in self.db_to_conns.items():
-                for conn in conns:
-                    try:
-                        conn.close()
-                    except:
-                        pass
-            self.db_to_conns.clear()
-
-
-class DBAffinityConnectionsNoLimit(object):
+                self.put_conns(conns)
+                conns = []
+                raise
 
-    # Similar to the db affinity pool, but without keeping track of
-    # the connections. It's the caller's responsibility to call us
-    # back with the connection objects so that we can close them.
-
-    def __init__(self, dbnames, n_conn_per_db, conn_info):
-        self.dbnames = dbnames
-        self.n_conn_per_db = n_conn_per_db
-        self.conn_info = conn_info
-        self.conn_mapping = {}
-        self.lock = threading.Lock()
-        self.dbname_index = 0
-
-    def _make_conn(self, conn_info):
-        conn = psycopg2.connect(**conn_info)
-        conn.set_session(readonly=True, autocommit=True)
-        register_hstore(conn)
-        register_json(conn)
-        return conn
-
-    def get_conns(self):
-        with self.lock:
-            dbname = self.dbnames[self.dbname_index]
-            self.dbname_index += 1
-            if self.dbname_index >= len(self.dbnames):
-                self.dbname_index = 0
-        conn_info_with_db = dict(self.conn_info, dbname=dbname)
-        conns = [self._make_conn(conn_info_with_db)
-                 for i in range(self.n_conn_per_db)]
-        # convenient for caller to get the conn_info for fetching the
-        # columns
-        return conns, conn_info_with_db
+        return conns
 
     def put_conns(self, conns):
-        for conn in conns:
-            try:
-                conn.close()
-            except:
-                pass
+        with self._lock:
+            for conn in conns:
+                pool = self._conns_to_pool.pop(id(conn), None)
+                if pool:
+                    pool.putconn(conn)
```

Comment:
I recognize that this case should never happen, but should we have an else clause where we try to close the connection? Actually an assert might make more sense because this means a caller tried to close a connection that it didn't get from a pool.